packages/graphql/lib/src/links/websocket_link/websocket_link.dart

```diff
@@ -39,6 +41,7 @@ class WebSocketLink extends Link {
     _socketClient = SocketClient(
       url,
       config: config,
+      protocol: subProtocol,
```

Comment:
For `graphql-transport-ws`, connect needs to block before subscribe is sent, otherwise a 4401 error will be returned.

```
  void connectOrReconnect() async {
    _socketClient?.dispose();
    _socketClient = await SocketClient(
      url,
      config: config,
      protocol: subProtocol,
    ).connect();
  }
```

Reply:
Mh I take a long reflection on this site and I think you have right! I usually don't use the lock when I write a client for the client because multiple requests can have multiple responses.

However, when we insert this logic inside a flutter app, this will be a little bit messy!

I will try to move this code with the lock, but I will take a little bit of time to reason on it. I'm 100% sure of this choice!

Reply:
Rather than using a mutex, you can probably just store the future (or maybe `FutureOr`) in `_socketClient` instead of the `SocketClient`.

```dart
Future<SocketClient>? _socketClient;
```

then you'd await that future in `request`.